### ISSUE TITLE
RFC: Major version bump - Kint 2.0-dev

### ISSUE THREAD

**Issue by jnvsor** [kint-php/kint#208]
Today, I've pushed something to [jnvsor/kint/2.x-dev](https://github.com/jnvsor/kint/tree/2.x-dev). I've been working on this for about 3 months now, I've basically rewritten all of Kint from the ground up.

I'd like this to be a starting point for a new major version of kint.
## Bugfixes
#50 #175 #203 #197 #182 #177 #176 And probably more too...
## Plugins

Lots of kint is very rigid. It's not designed with extensibility in mind. Most of the time you want to customize behaviour you have to edit kint source code yourself. In some of the best cases you have to dump new files inside kint's directory tree. These are bad things for anyone that wants to install an update.

My branch has a very extensive plugin system, to the point where you can pretty much replace both the parsing and rendering systems entirely without ever touching kint source code. This makes it possible for user provided solutions to certain feature requests:
#107 (High prio for over 2 years! :D) #185 #94 #189
#87 : No longer applies: All the config system did was set a few properties on the kint class, you can do that from anywhere and have an easier time of it and a cleaner codebase

Additionally, because the parser/renderer/plugins etc are properly decoupled, it can be meaningfully _unit tested_. (This provides a start at #26)

In fact, it's so decoupled that if you want you can just call the parser and renderer and leave the main Kint class out of the loop entirely. (TODO: Some plugins get hardcoded info from Kint class, `max_str_length` in `Kint_Object_Blob` for example)
## Internals

Kint master internals are Lovecraftian - peer at them too long and your sanity will be devoured by the great old ones.

Sarcasm aside - kint seems to have been developed very organically, which isn't necessarily bad but tends to add technical debt.
- We have a parser that outputs HTML, so if you want a decorator that doesn't you first have to unparse the HTML (See 2db0845f7a9ac642d09947e6f000e57e3b91ac64 )
- For some reason the parser is actually a child class of the data objects it's supposed to produce (???)
- The parser tries valliantly (And fails miserably) to prevent recursion "ghosts" on referenced arrays at the root level (See #203, #204) I've explained why this can't be done before, but that's a complex topic so I'll leave that for the other issues.
- Parser plugins almost exclusively produce html, so they're by definition reliant on the rich renderer. Checks on `Kint::$enabledMode` are strewn throught the parser system.
- It's a deck of cards. Everything is far too tightly coupled.

Long story short: I gutted everything. Now the parser just returns data, and the renderer takes that data and does whatever with it. The individual parts are all smaller: The rich renderer is 10% smaller, the parser is 40% smaller, the JS renderer (The example for the unparsing problem) is also 40% smaller.

Lots of special case code has been moved from the parser and main kint class into plugins where they belong: Backtrace handling, tabular display, etc

Basically it's a big-ass cleanup and for the most part easier to grep
## The catch

Well not catch so much as caveats.
### This is a work in progress

While it works fine for basic debugging, most of the plugins aren't ported yet. `ClassStatics`, `ClassMethods`, `Microtime`, `Closure` and `Trace` have all been ported, but there's still a small mountain worth of missing ones (Including things like the tabular display that used to be hardcoded in the parser)

The plain renderer also hasn't been ported yet, so `PLAIN` `WHITESPACE` and `CLI` output modes won't work till that's done.

It's got a lot of good stuff, but it's not feature parity (Yet)
### This is not a stable API

I'm very happy with the way the parser works. I've put a lot of thought into it and it seems to work beautifully. The problem is the renderer.

The renderer needs to know how to take data and turn it into output. Currently, the parser plugins can add a "Class" to a representation (Tab) and the renderer maps that class to a renderer plugin. This works fine but there's no way to alter the way a _header_ is rendered, only a tab. I intend to make large changes to the plugin system until it's working well enough.

Here be monsters.
### It's slow

Last I checked the parser was some 1000% faster than the old parser, but the renderer is significantly slower and it all evens out at a 50% slowdown (Last I checked... I've changed a lot since then)

_I'm not actually sure why this is_ - when I stub out all the relevant functions so it's just outputting enough to let me know it's actually outputting at all, it's still far slower. I'm guessing it's some PHP internal optimization - probably something in static methods that PHP deems pure so it caches the result or some such.

All that said, it's not too slow - I don't see any difference in normal use. The pliability benefits of the new plugin system outweigh a speed difference I can't tell exists without tools. I'll take a closer look after I'm satisfied with the API.
### It works in PHP 5.1

While I'd like to make a smiley and say this is just a joke about kint's strict version requirements (next month 5.1 will have been EOL for a decade) the truth is _it's so old I couldn't find a way to get it working_ so while it absolutely should work on 5.1 I have **NO** way of proving that. If anyone can test it for me (Or show me an easy way to do it myself) that would be a big help.
### The code style

Code style conversations are always purely subjective. Subjective enough to turn into flame wars so I'm not going to bother rooting for the code style I use here, or against the code style in master.

Instead I'm going to argue on consistency. Consistent code style is good right? We can agree on that right? Right?...

Anyway, I'm not very consistent. I drift between K&R, PSR, trololol-minified-php-ftw, and generally don't have anything consistent enough to write my own code style rules. So I let my computer do it for me.

[php-cs-fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) guarantees consistent code. It also lets me write whatever sloppy nonsense I want to then fix it with a simple command. Because it's a program, we could also enforce it alongside a test suite with CI.

There are other purported php formatters out there but of the few I've tried php-cs-fixer is the best (One of them actually changed method naming styles and broke a system on me once... Yay) and the defaults are fairly sane and PSR compliant. Most importantly, I personally can attest that if I stuck with master's code style I'd have been at it for another 3 months :)
## What needs to be done

If anyone has any ideas that could clean up the architecture I've made here that would be _great_
- I want to tweak the rendering system a bit so that plugins can effect headers as well as tabs, so I'll probably end up gutting that for the fifth time.
- It needs to be tested for 5.1 compat
- I need to find out what's going on with rendering performance. This should wait until all plugins are ported as that can effect performance too.
- A lot of plugins need to be ported
- A test suite would be nice, but lets get things working first.
- I want to decouple a few things that shouldn't be in the main Kint class (max_str_length should probably be on the renderer not the Kint class, the Trace plugin should have a configurable alias tracking system, etc)

**Comment by ATouhou** [comment]
Nice! Good job! :D 
Works with php7?

**Comment by jnvsor** [comment]
Yep, I've been switching back and forth, last I checked I was developing on 7 (But master also works on 7 so that's not that big a deal ;P)

**Comment by xanthakita** [comment]
Man sounds extremely impressive!

**Comment by raveren** [comment]
Hello, I am sorry for the absence, I am preparing a code review for this, I just have zero time, thanks for this tremendous effort and I will come back to this soon!

**Comment by raveren** [comment]
Ok, I took a look and I can't get it to actually work:

``` php
<?php

require 'k/init.php';
$a = [ "Array a" ];
$b = [ "Array b" ];
$a[] = &$b;
$b[] = &$a;
$c = [ &$b ];
ddd( $a, $b, $c );
```

just throws `Catchable fatal error: Argument 6 passed to Kint_Renderer_Rich::__construct() must be of the type array, boolean given, called in <redacted>/k/src/Kint.php on line 241 and defined in <redacted>/k/src/Renderer/Rich.php on line 22`.

Also, I don't really appreciate your take on my work and since I now feel under attack I'm going to address some of your points, for example in your chapter **Internals**:

>    We have a parser that outputs HTML, so if you want a decorator that doesn't you first have to unparse the HTML (See 2db0845 )
- I've experimented with different renderers for countless hours through out the years and the current situation is what I came to be happy with: you either get lots of extra info in **html** format or just the basics in plain. I _could_ be persuaded otherwise, but the real problem we're solving is: internal framework objects that are too big to display and I don't think your approach is the best for that.

>    For some reason the parser is actually a child class of the data objects it's supposed to produce (???)
- That is made so to prevent public methods from being available where they shouldn't be. I wish this could have been more elegant, but PHP is rigid in that regard. The current solution did the purpose at the time of writing, I could perhaps come up with something better now, and I think you might have done that (I'm withholding a real review until I get your version to work at all).

>    The parser tries valliantly (And fails miserably) to prevent recursion "ghosts" on referenced arrays at the root level (See #203, #204) I've explained why this can't be done before, but that's a complex topic so I'll leave that for the other issues.
- I'm not even going to discuss this point with you again, you didn't listen what I had to say and I'm not going to repeat myself.

>    Parser plugins almost exclusively produce html, so they're by definition reliant on the rich renderer. Checks on Kint::$enabledMode are strewn throught the parser system.
- Duplicated argument, see first point.

>    It's a deck of cards. Everything is far too tightly coupled.
- Well yes, kint would not suffer from a rewrite, but it was the last thing on my mind, even though I'm hoping, that your changes would make further contributions easier which is a real goal. 

If your changes (evidently not in their current state, but eventually) make Kint easier to extend and contribute to, I will move the project under a github organization and make you a co-owner.

**Comment by jnvsor** [comment]
```
Catchable fatal error: Argument 6 passed to Kint_Renderer_Rich::__construct() must be of the type array, boolean given, called in <redacted>/k/src/Kint.php on line 241 and defined in <redacted>/k/src/Renderer/Rich.php on line 22
```

Ok, nice spot. Forgot that could happen - pushed a bugfix just now

> you either get lots of extra info in html format or just the basics in plain

I was considering completely different output formats: Js, PDF, SQL, some obscure binary format someone wants to write to, so I think there's use in having it

> internal framework objects that are too big to display and I don't think your approach is the best for that

Can you give me an outline of your thoughts on this?

**Comment by raveren** [comment]
I'm very against feature creep, as I outlined in the contribution guidelines, too many unneeded stuff gets left behind and just weigh projects down, so I'd very much prefer to not even include the js output, in all of my years I've never saw a need for something like this. Kint once had a feature-complete ajax output mode with a modular window that would display if anything was dumped inside an ajax call. It cost me weeks to write but eventually I just threw it out without releasing publicly cause my colleagues (and unofficial beta testers:) were more confused by it than anything.

It's better to force users to be creative than weight the project down, unless their usecase is not something extremely exotic **and** cannot be solved with the existing tools, only then you should implement something.

As for the framework object what I meant was that a real concern is huge framework and 3rd party objects (like the smarty one that is handled) that hang browsers without mercy when accidentally dumped alongside real info or in traces. To understand what I mean just dump a trace in a Zend framework project...

**Comment by jnvsor** [comment]
> so I'd very much prefer to not even include the js output

It's only 50 lines, but you're right and I'd be fine releasing it as a separate plugin. It does however provide useful feedback for the renderer API so I'd like to leave it in until I'm satisfied with that side of things

> To understand what I mean just dump a trace in a Zend framework project...

Yeah - especially through a templating engine! Though I still think this should be less about the renderer perse and more about the plugins (I'm sure your zend trace for example would be leaner if all the plugins were disabled)

This sort of reminds me of #185 - and the new parser plugin system makes it fairly easy for end users to exclude framework objects if needs be. You could just add a plugin with something like...

``` php
<?php

class DontHangMeBrowserMatey extends Kint_Parser_Plugin
{
    public function parse(&$var, Kint_Object &$o)
    {
        if (typeof($var) == 'object' && substr(get_class($var), 0, 4) == 'Zend') {
            $o = new NotHangingYourBrowserMatey($o);
        }
    }
}

class NotHangingYourBrowserMatey extends Kint_Object_Instance
{
    public function __construct($o)
    {
        $o->transplant($this);
        $this->value_representation = null;
        $this->representations = [];
    }
}
```

**Edit:** Now that I think of it, perhaps bundling a `Blacklist` plugin for common frameworks might be an idea for later?

**Comment by raveren** [comment]
@jnvsor 

> Ok, nice spot. Forgot that could happen - pushed a bugfix just now

um there's no bugfix, the recent code is the same and throws the same error...

**Comment by jnvsor** [comment]
Oh shit sorry, must have pushed the wrong branch... It's there now

**Comment by raveren** [comment]
Hello, unfortunately, as far as functionality goes, most of the things I try result in a regression. I couldn't force alternative display to show (I tried dumping json and `time()`), and I see inconsistency in the `src\Parser\Plugin\*` files. Trace is broken as well, no more `d(1)` shorthand, and you discarded my work on early recursion detection (although mine indeed needs improvement, but not be thrown out): 

``` php
<?php

require 'k/init.php';
$a = [ "Array a" ];
$b = [ "Array b" ];
$a[] = &$b;
$b[] = &$a;
$c = [ &$b ];
ddd( $a, $b, $c );
```

![image](https://cloud.githubusercontent.com/assets/364220/17620228/7d526274-6093-11e6-8f4c-fee1153b55a0.png)
I could have tested further, but this all can be fixed in time and does not concern me.

What I wanted to do is give you comments as you requested, but I don't see a clear direction you are taking the project. As I have stated somewhere already, the clear broad goals that I can currently see are: easier contribution and external configuration - and while refactoring _might_ be a viable approach, I am not yet convinced. 

To reiterate, I can see a lot of careful work has been put into this, but no clear direction and goals are defined. The effort seems to have been done to scratch a personal itch, when in fact you need a list of _real, hard, actually needed_ requirements that your work needs to fulfill. Otherwise after all the effort it might still be debatable which code base is objectively better - and in that situation I always vote for the field-tested code.

Another unfortunate thing, I am leaving tomorrow for a whole month, we will only be able to continue after I return..

**Comment by raveren** [comment]
P.S. I love your work on #175, you could pre-select the text immediately after revealing it. You could also create a keyboard shortcut to access the functionality if there's a completionist in you, but even now it's impressive and useful!

**Comment by jnvsor** [comment]
> Hello, unfortunately, as far as functionality goes, most of the things I try result in a regression

I did mention it's not at feature parity yet - the only plugins currently enabled are microtime, objectstatics, objectmethods, closures, and backtrace. The inconsistency is because most of the plugin files are pretty much verbatim the old code and I haven't touched them yet.

> Trace is broken as well, no more d(1) shorthand

This is related to the same problem you had earlier - `getCalleeInfo` is having trouble finding the correct spot in your trace. In my tests it worked fine, and I couldn't reproduce your situation (I forgot it could ever happen, hence the bug you ran into testing it the first time)

It might be an off-by-one if you're running kint from the global scope - I'll check that this weekend.

> you discarded my work on early recursion detection

I'll continue this discussion in #204 - I'm still 99(.9)% sure your desired behaviour isn't possible, and I'll try to explain my reasoning again

> you need a list of real, hard, actually needed requirements that your work needs to fulfill

As you mentioned, easier contribution is there, but that wasn't so much a goal as something that happened automatically because I found it hard to contribute myself - CI would be a bigger help in this area (Even before a working test suite it could be used to enforce CSS/JS compilation and code style) and given @voku 's work in #205 I think he's probably the most qualified to help set that up.

External configuration is a real goal though: Kint is great but I find myself reluctant to use it in real-life situations because I have to maintain a separate branch for any changes, even to something as simple as CSS.

The goal for the plugin system, while similar to configuration, was the main reason for this rewrite. Half the issues this repo receives are feature requests that don't get implemented due to being too specific.

While kint master has a plugin system of sorts, it's addled by the lack of renderer configuration and the fact you have to drop the files inside the kint directory - another reason for the external config.

If I were to name goals for the whole thing I'd probably list:
- Ease of contribution via code cleanup
- Ability to change behaviour from outside the kint folder
- A more flexible parser/renderer/plugin system that allows complete output customization
- Feature parity with kint master (As mentioned before - a work in progress)
- A few new features (Notably access path) I've determined since starting this branch that proper access path behaviour can't be achieved without changes to the parser and main kint class on a large enough scale that the parser and main kint class would need to be rewritten anyway

> P.S. I love your work on #175

That reminds me, don't merge any of my past PRs! They're all in this branch in an improved form. For example, backtrace paths will be completely wrong in #175 (aka #177) because I didn't know `_parseTrace` removes certain frames. Same goes for #197 #182 #176

Regarding #182 - 2.x kint has an extra option to delay output to the end of a script by means of `register_shutdown_function` but I've been thinking this might as well be taken out of the Kint class and left in the helper functions - do you have thoughts on this before you go on vacation?

**Comment by jnvsor** [comment]
Since @raveren will probably be back soon, here's a list of questions I've got for you
- Would you mind if 2.x-dev switches from closure to uglify-js for javascript compression? It removes a Java dependancy and is far faster at building. It does less optimisation but kint's JS isn't the heaviest stuff out there anyway
- I've experimented with TravisCI on one of my other projects in the meantime - would you like me to write preliminary tests for at least code style? How exactly should I go about it? I imagined just slapping PHPunit on it but PHPunit hasn't supported PHP 5.1 in [half a decade](https://github.com/sebastianbergmann/phpunit/releases/tag/3.5.15)
- Should we be keeping compiled CSS/JS in the repo, or just compile them into release tarballs and have more regular releases?
- Should my `register_shutdown_function` implementation go inside the helper function or inside kint itself? (It's simple enough to implement so I don't think there's any harm in moving it to the helper function)
- I still can't reproduce your minitrace issue. Could you send me a `var_dump(debug_backtrace())` from inside `getCalleeInfo`?

**Comment by raveren** [comment]
I'll quickly answer your questions while I have a minute. In order:

1) if uglify makes development easier (which I believe it does), I don't mind phasing out Closure compilation, it gives no real benefit to the user.
2) I am not dropping compatibility with older PHP versions just because it's more difficult to develop - unless end-users are actually affected. I don't like tools that enforce a code style either. Unless the project is developed by a big, ranging in seniority and rapidly changing team there should be no need for it, just another tool in the development stack that adds no real value. You commit only the lines you actually change and adapt the code style from the rest of the file/project, don't overengineer.
3) I think we should keep them. Kint boasts effortless setup, if you download it, you must be able to run it.
4) Does your functionality solve a highly requested workflow or is it just something that adds complexity to the project and will seldom see use? 
5) What minitrace issue? I didn't try your RFC after you fixed it yet...

**Comment by jnvsor** [comment]
Welcome back!

> I am not dropping compatibility with older PHP versions just because it's more difficult to develop

Ok, we'll probably have to either build our own test suite or add an old phpunit as a submodule if we want tests.

> I don't like tools that enforce a code style either.

The code readability goal in the contrib file would benefit from it, and last I checked you went over PRs manually and pointed out style issues anyway

> I think we should keep them. Kint boasts effortless setup, if you download it, you must be able to run it.

Github has the ability to upload zip files with releases (Tags) which has been the standard way to include precompiled stuff from source in other projects, but that works too. We'll have to add a test to enforce compilation on changes though

> Does your functionality solve a highly requested workflow or is it just something that adds complexity to the project and will seldom see use?

Well you requested it way back in #50 

> What minitrace issue?

Basically, something's wrong in the main kint class that's causing a few issues on your end (Missing minitrace I see in the screenshots, and the uncaught exception from [earlier](https://github.com/raveren/kint/issues/208#issuecomment-237484365))

**Comment by raveren** [comment]
Hey! 

> Ok, we'll probably have to either build our own test suite or add an old phpunit as a submodule if we want tests.

Tests are in a way just a measure to ease development. However I've been thinking and if I could separate a bare-bones but still objectively useful version of Kint that would support legacy versions (down to say PHP4.4), we could lift the requirements to a reasonable PHP version for the main fork.

> The code readability goal in the contrib file would benefit from it, and last I checked you went over PRs manually and pointed out style issues anyway

I never said style or concistency is unimportant, I just wouldn't want to introduce a tool that I don't see a real need for, but I can reconsider if you find it useful

> Github has the ability to upload zip files with releases (Tags) which has been the standard way to include precompiled stuff from source in other projects, but that works too. We'll have to add a test to enforce compilation on changes though

Github also has the Download ZIP feature that is more readily available and most probably used a lot. It fetches the code in the HEAD revision.

> Well you requested it way back in #50

I migrated this issue from google code. It seeing minimal activity since 2013 speaks for itself. If you read the contribution guidelines, I'm very against feature creep and I just don't see this functionality as useful. If you give me a scenario where this achieves a result better than pre-existing solutions, I will reconsider, but for now I doubt this needs implementation at all..

> Basically, something's wrong in the main kint class that's causing a few issues on your end (Missing minitrace I see in the screenshots, and the uncaught exception from earlier)

The minitrace is there, the dump was called in top-level, that's it. I haven't tried out your newer changes yet, that were supposed to solve the exception...

**Comment by jnvsor** [comment]
> if I could separate a bare-bones but still objectively useful version of Kint that would support legacy versions (down to say PHP4.4), we could lift the requirements to a reasonable PHP version for the main fork.

At that point it would be easier just to keep kint supporting 5.1

> but I can reconsider if you find it useful

I do, but it needs to be either or - if we don't adhere to it strictly then all commits after a breaking one would complain, so either we keep to it or don't use it at all

> Github also has the Download ZIP feature

There's also the option of having CI build it for us and stick a link to that in the readme

> It seeing minimal activity since 2013 speaks for itself

Alrighty, I'll remove this from my branch. If you look at the issues queue there are a lot of old ones like that (#40 for example exists already with double right arrow)

**Comment by jnvsor** [comment]
> P.S. I love your work on #175, you could pre-select the text immediately after revealing it. You could also create a keyboard shortcut to access the functionality if there's a completionist in you, but even now it's impressive and useful!

Done both - picked `s` for the key for no particular reason - that ok with you? (Note that it won't work on leaf nodes since you can't select them with the keyboard :cry:)

**Comment by jnvsor** [comment]
Latest push includes
- Highlight access path on show
- Access path keybind (a)
- Bugfix for expanding minitrace with keybinds (Doesn't work in master)
- FsPath implemented (Also fixes buggy symlink detection - on my system at least)
- Json implemented (Check out the wonderous access_path magic here!)

![](http://i.imgur.com/yzkEYTl.png)

**Comment by jnvsor** [comment]
Latest push includes (For anyone still reading this)
- Microtime plugin now has a `clean()` method that will allow you to reset the lap counters to benchmark multiple loops
- **Added the color plugin** - this was a big one and took a lot of work.
  
  I decided to make the output look fairly different from the original.
  
  Key thing to note here is that the only changes to the parser/renderer were to add the plugins to the list. This can be done from outside code. So the entire color plugin from parsing to rendering can be implemented without touching any Kint code or directories.
  
  That's pretty sweet! :smile: 
  
  ![firefox_screenshot_2016-11-13t16-22-52 129z](https://cloud.githubusercontent.com/assets/5260644/20247179/ddc1926c-a9c5-11e6-8781-4e1a14042d9e.png)

**Comment by jnvsor** [comment]
Changelist:

* Rewrote the JS
* You can now use keyboard and mouse navigation at the same time
* Added a `kint-locked` class for objects that have children we can't show (Like depth limit and recursion)
* To support these I needed to add an icon to the caret so I went ahead and remade the caret spritesheets as handcrafted SVG - they all got more detail and smaller filesize (Yay)
* The rich renderer now has an array of callbacks it will use to insert JS and CSS into the page on first run which can be useful for plugins. If you want to insert JS on a second run you're out of luck for now
* Simplified the rich plugin system a bit
* Added automated codestyle fixers for PHP SCSS and JS. SCSS is fairly simple, PHP is fairly complex.
* Merged master with 2.x-dev so it's a fast-forward from raveren/master now
* Changed source code representation to allow copy paste without line numbers
* I fixed #218 back in e6c57edd16682e but I thought I'd mention it here
* Bugfixes for Color, TraceFrame, Json, Parser, Rich Renderer, 
* Added timestamp plugin
* Added Iterator plugin, with blacklist for specific classes (Fixes #217)
* Added Blacklist plugin (To completely block debug of specific objects - useful for sensitive info like passwords. Fixes #185 (And to an extent #186, #189)
* Added SimpleXMLElement parsing plugin
* Added DOMDocument parsing plugin (This was a 3 week pain in the ass)
* Added XML plugin. Configurable to use either SimpleXML or DOMDocument, but it uses SimpleXML by default since DOMDocument is slower and uses more memory
* Simplified the representation API in objects. I'm pretty happy with objects now so I doubt they'll change much before release.
* Added Serialize plugin
* Added Base64 plugin
* Updated contributing.md and composer.json for Kint-PHP
* Removed Js renderer (It'll be its own repo)
* Removed a number of helper functions that amounted to `d();exit;`
* Added tabular data plugin (Simpler than the old one, but less inconsistant IMHO in that it doesn't allow an arbitrary amount of missing elements)
* Pushed to Kint-PHP (Yay?)

**Comment by cdillon** [comment]
I want to integrate Kint with WordPress; specifically, a config file outside of the Kint directory. Will version 2.0 make this simpler?

**Comment by jnvsor** [comment]
Yep. You can change pretty much the entire debugging process without touching the kint dir.

That said the plain renderer isn't finished yet and I've only just got a VM working with PHP 5.1 so it'll probably throw errors below 5.6 until I push the compat changes (Should be done today I hope)

Check kint-php/kint#1 for the current todo list

That said: All the kint config file does in 1.x is set a bunch of statics on the Kint class, so you can probably integrate 1.x with little difficulty (Make sure to use master since it hasn't been tagged in a while)

**Comment by jnvsor** [comment]
I'm now in charge of the Kint repository and the [2.x-dev](/kint-php/kint/tree/2.x-dev) is available for feedback

**Comment by jnvsor** [comment]
Kint 2.0 is now in beta. No more architectural changes, feedback and bug reports should be new issues.
